### PR TITLE
k-nearest-neighbor-connected  ground meshes for TrussTO 

### PR DESCRIPTION
**Pull request by yijiangh**
### Summary

This PR includes a new problem type for truss topology optimization called PointLoadCantileverTruss that creates equally spaced nodes in a rectilinear domain and connects linear elements between each node and its `k` nearest nodes. We have both 2D and 3D supports.

2D 1-nearest-neighbor-connected ground mesh and its compliance-minimized result:
image
image

2D 2-nearest-neighbor-connected ground mesh:
image

3D 1-nn:
image

**Minor changes**:
1. Add `NearestNeighbors` dependencies.

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

- [x] I ran all tests on my computer and it's all green (i.e. `] test`).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -16,6 +16,9 @@ GeometryTypes = "4d00f742-c7ba-57c2-abde-4428a4b178cb"
 IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
 JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+MacroTools = "1914dd2f-81c6-5fcd-8719-6d5c9610ff09"
```
Comment (mohamed82008) [inline Project.toml:19]: Please add a compat line for each new dependency

Comment (mohamed82008) [reply Project.toml:19]: This is required by JuliaRegistrator

Comment (yijiangh) [reply Project.toml:19]: Done!

**Comment by mohamed82008** [comment]
Great PR as always!

**Pull request merged**